src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+
+lemma has_groupoid_of_le {G₁ G₂ : structure_groupoid H} (h : has_groupoid H M G₁) (hle : G₁ ≤ G₂) :
+  has_groupoid H M G₂ :=
+⟨ λ e e' he he', hle ((h.compatible : _) e e' he he') ⟩
+
+/-- The trivial manifold structure on the model space is compatible with any groupoid -/
+instance has_groupoid_model_space (H : Type*) [topological_space H] (G : structure_groupoid H) :
+  has_groupoid H H G :=
+{ compatible := λe e' he he', begin
+    replace he : e ∈ atlas H H := he,
+    replace he' : e' ∈ atlas H H := he',
+    rw model_space_atlas at he he',
+    simp [he, he', structure_groupoid.id_mem]
+  end }
+
+/-- Any manifold structure is compatible with the groupoid of all local homeomorphisms -/
+instance has_groupoid_continuous_groupoid : has_groupoid H M (continuous_groupoid H) :=
+⟨begin
+  assume e e' he he',
+  rw [continuous_groupoid, mem_groupoid_of_pregroupoid],
+  simp only [and_self]
+end⟩
+
+/-- In a manifold having some structure groupoid, the changes of coordinates belong to this groupoid -/
+lemma compatible (G : structure_groupoid H) [h : has_groupoid H M G]
+  {e e' : local_homeomorph M H} (he : e ∈ atlas H M) (he' : e' ∈ atlas H M) :
+  e.symm ≫ₕ e' ∈ G :=
+(h.compatible : _) e e' he he'
+
+/-- A G-diffeomorphism between two manifolds is a homeomorphism which, when read in the charts,
+belongs to G. -/
+structure diffeomorph (G : structure_groupoid H) (M : Type*) (M' : Type*)
+  [topological_space M] [topological_space M'] [manifold H M] [manifold H M']
+  extends homeomorph M M' :=
+(smooth_to_fun : ∀c : local_homeomorph M H, ∀c' : local_homeomorph M' H,
+  c ∈ atlas H M → c' ∈ atlas H M' → c.symm ≫ₕ to_homeomorph.to_local_homeomorph ≫ₕ c' ∈ G)
+
+variables [topological_space M'] [topological_space M'']
+{G : structure_groupoid H} [manifold H M'] [manifold H M'']
+
+/-- The identity is a diffeomorphism of any manifold, for any groupoid. -/
+def diffeomorph.refl (M : Type*) [topological_space M] [manifold H M]
+  [has_groupoid H M G] : diffeomorph G M M :=
+{ smooth_to_fun := λc c' hc hc', begin
+    change (local_homeomorph.symm c) ≫ₕ (local_homeomorph.refl M) ≫ₕ c' ∈ G,
+    rw local_homeomorph.refl_trans,
+    exact compatible G hc hc'
+  end,
+  ..homeomorph.refl M }
+
+/-- The inverse of a diffeomorphism is a diffeomorphism -/
+def diffeomorph.symm (e : diffeomorph G M M') : diffeomorph G M' M :=
+{ smooth_to_fun := begin
+    assume c c' hc hc',
+    have : (c'.symm ≫ₕ e.to_homeomorph.to_local_homeomorph ≫ₕ c).symm ∈ G :=
+      G.inv _ (e.smooth_to_fun c' c hc' hc),
+    simp at this,
+    rwa [trans_symm_eq_symm_trans_symm, trans_symm_eq_symm_trans_symm, symm_symm, trans_assoc]
+      at this,
+  end,
+  ..e.to_homeomorph.symm}
+
+/-- The composition of diffeomorphisms is a diffeomorphism -/
+def diffeomorph.trans (e : diffeomorph G M M') (e' : diffeomorph G M' M'') : diffeomorph G M M'' :=
+{ smooth_to_fun := begin
+    /- Let c and c' be two charts in M and M''. We want to show that e' ∘ e is smooth in these
```

Comment:
Again, the word "smooth" is utterly confusing in such a general context.

Reply:
What about `to_fun_mem_groupoid` (which is slightly misleading, since `to_fun` belongs to the groupoid only when read in charts, but hey).

Reply:
Half serious: `structo_to_fun`